pkg/signatures/digester.go

```diff
@@ -0,0 +1,162 @@
+	}
+	return &cdv2.DigestSpec{
+		HashAlgorithm:          d.hasher.AlgorithmName,
+		NormalisationAlgorithm: string(cdv2.LocalOciBlobDigestV1),
+		Value:                  hex.EncodeToString((d.hasher.HashFunction.Sum(nil))),
+	}, nil
+}
+
+func (d *Digester) digestForOciArtifact(ctx context.Context, componentDescriptor cdv2.ComponentDescriptor, res cdv2.Resource) (*cdv2.DigestSpec, error) {
+	if res.Access.GetType() != cdv2.OCIRegistryType {
+		return nil, fmt.Errorf("unsupported access type: %s", res.Access.Type)
+	}
+
+	ociAccess := &cdv2.OCIRegistryAccess{}
+	if err := res.Access.DecodeInto(ociAccess); err != nil {
+		return nil, fmt.Errorf("unable to decode resource access: %w", err)
+	}
+
+	ociArtifact, err := d.ociClient.GetOCIArtifact(ctx, ociAccess.ImageReference)
+	if err != nil {
+		return nil, fmt.Errorf("failed resolving oci artifact: %w", err)
+	}
+
+	bytes, err := ociArtifact.Raw()
+	if err != nil {
+		return nil, fmt.Errorf("failed getting oci artifact raw data: %w", err)
+	}
+
+	d.hasher.HashFunction.Reset()
+	if _, err = d.hasher.HashFunction.Write(bytes); err != nil {
+		return nil, fmt.Errorf("failed hashing yaml, %w", err)
+	}
+
+	return &cdv2.DigestSpec{
+		HashAlgorithm:          d.hasher.AlgorithmName,
+		NormalisationAlgorithm: string(cdv2.ManifestDigestV1),
+		Value:                  hex.EncodeToString((d.hasher.HashFunction.Sum(nil))),
+	}, nil
+}
+
+func (d *Digester) digestForS3Access(ctx context.Context, componentDescriptor cdv2.ComponentDescriptor, res cdv2.Resource) (*cdv2.DigestSpec, error) {
+	log := logger.Log.WithValues("componentDescriptor", componentDescriptor.ComponentSpec.ObjectMeta, "resource.name", res.Name, "resource.version", res.Version, "resource.extraIdentity", res.ExtraIdentity)
+
+	if res.Access.GetType() != cdv2.S3AccessType {
+		return nil, fmt.Errorf("unsupported access type for s3 Access Digester: %s", res.Access.Type)
+	}
+	s3Access := &cdv2.S3Access{}
+	if err := res.Access.DecodeInto(s3Access); err != nil {
+		return nil, fmt.Errorf("unable to decode resource access: %w", err)
+	}
+
+	url := fmt.Sprintf("https://%s.s3.amazonaws.com/%s", s3Access.BucketName, s3Access.ObjectKey)
+	resp, err := http.Get(url)
+	if err != nil {
+		return nil, fmt.Errorf("unable to access s3 access with url %s: %w", url, err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("unable to access s3 access with url %s, response code %d", url, resp.StatusCode)
+	}
+	log.V(5).Info(fmt.Sprintf("downloading and hashing %s bytes from s3 access", resp.Header.Get("Content-Length")))
+	d.hasher.HashFunction.Reset()
+	if _, err := io.Copy(d.hasher.HashFunction, resp.Body); err != nil {
+		return nil, fmt.Errorf("unable to hash s3 access with url %s and hash function %s: %w", url, d.hasher.AlgorithmName, err)
+	}
+	return &cdv2.DigestSpec{
+		HashAlgorithm:          d.hasher.AlgorithmName,
+		NormalisationAlgorithm: string(cdv2.GenericBlobDigestV1),
```

Comment:
Isn't this the same as cdv2.LocalOciBlobDigestV1? we only download the byte stream and calculate hash for this

Reply:
This is correct. But I would use the GenericBlobDigestV1 instead of LocalOciBlobDigest for localOciBlobs since it is more generic.